### DISCUSSION TITLE
Support for TUnit

### DISCUSSION THREAD

**Discussion by dev-in-disguise** [AutoFixture/AutoFixture#1478] [Ideas]
A new unit testing framework exists that's called [TUnit](https://github.com/thomhurst/TUnit). 

Personally I'd find it awesome if AutoFixture would have support for this framework as well since it would make my migration path easier. 

Would it be acceptable if I or someone else would want to add support for the several attributes that need unit testing framework specific implementations like `AutoData`? 
And of course would PRs for this also be approved since I'm aware that more supported frameworks add more maintenance effort for the library itself in the future?

**Comment by Kralizek** [comment]
I agree that support for TUnit would be really interesting but I hardly it can be fit into v5.

**Comment by Ergamon** [comment]
I took a look at the project. So far I have not found a good way to realize what we have in xUnit so far.
It is easy to inject AutoData via ctor with TUnit. Doing the same in xUnit is a pain in the ass. (I use my own version of AutoFixture.xUnit.

Sadly I have not found anything in TUnit for having AutoData like attributes. You can use MethodDataSource to inject AutoData. But the method has no clue about the argument types.

The author has currently a discussion about lacking features. I made a short post. Lets wait if something usable comes out.

**Reply by Kralizek** [reply]
Can you link the post or thread in your message above?

**Reply by Ergamon** [reply]
No, it is top secret ;)

https://github.com/thomhurst/TUnit/discussions/646

**Reply by dev-in-disguise** [reply]
Apart from the discussion and potential enhancement in TUnit theoretically it would still be possible to have a parameterless AutoData attribute like it is currently. 

Let me first make it visible how I would envision the usage of the idea that I have to make it easier to explain:

```
public partial class TestFixture
{
    [AutoData]
    [Test]
    public async Task MyTest(int myInt, string myString)
    { ...Test Implementation }

    public static partial IEnumerable<(int, string)> AutoDataSource_MyTest();
}
```

As you potentially already can guess the idea involves source generators. Specifically there would be 2 source generators plus for easier usability one analyzer/code fixer. 

The first source generator would enable us to support the parameterless AutoData attribute. The AutoData attribute would inherit from TUnits `MethodDataSourceAttribute` and to be able to give it a correct method name as data source the source generator would implement a method that returns a unique name. The unique name should be based on the test method name plus a prefix. In the usage example it would return `AutoDataSource_MyTest`.
In theory we should also have a variant of the `AuoDataAttribute` which allows passing a type to support the second overload of TUnits `MethodDataSourceAttribute` but that doesn't change a lot in the concept. 

The second source generator has to implement the method data source. So in the usage example the `AutoDataSource_MyTest` method. I'm not too deep into the detailed implementation of AutoFixture but I guess this would basically implement what currently the `GetData` method does in the `AutoFixture` attribute implementations of other frameworks.

The additional analyzer/fixer portion would look for the `AutoDataAttribute` on test methods and add an error if a partial method definition for the data source does not exist. And the fixer would enable a user to generate the partial method definition. 

Excuse me please for this rough sketch but I hope this helps understand what I currently think about. 

Are there any concerns from your points of view? Is this in any way feasible and also something that you would like to have in such a  way? It surely is a completely different way of handling things compared to other testing frameworks out there.

**Reply by thomhurst** [reply]
Hey guys. Author of TUnit here, let me know if I can help in any way.

I think the easiest way is to create an attribute inheriting from NonTypedDataSourceGeneratorAttribute.

This returns just an object array.

You're also passed an object containing details about the parameters, including things like attributes on them, so I think that contains everything you might need?

The MatrixDataSource is built off of the same functionality, so source code for an idea of how it is set up is here: https://github.com/thomhurst/TUnit/blob/main/TUnit.Core/Attributes/TestData/MatrixDataSourceAttribute.cs

**Comment by dev-in-disguise** [comment]
Thanks that you're all already looking into it. I'm trying to play around a bit this week as well on a fork to check how feasible it is to make the implementation for TUnit. 

> I agree that support for TUnit would be really interesting but I hardly it can be fit into v5.

Personally that's no big deal for me as I'm just toying around with TUnit to potentially use it in my company. And this is why I stumbled upon the fact that Autofixture does not support it so far. So really no timing issue for me. 
Still it would be nice if we can ship support at some point in the future since at least the performance gains look really promising for this Framework so it could be that it gains in popularity over time.

**Comment by ahmed-abdelrazek** [comment]
any update on that feature?

**Comment by Ergamon** [comment]
I have some working code. It is at the current state more a conversion of the existing xunit code.

The question is basically where to put the code for further development. Should it be added here?

As written it is more a proof of concept, starting to use some of TUnits new possibilities.

Currently it lacks documentation and tests. As we I use NCrunch as test runner and NCrunch sadly does not support TUnit. There is a feature request, but it doesnt look good at the moment.
So most of the code I wrote out of curiosity. Without NCrunch support, I cannot use it in our projects.

**Comment by aivascu** [comment]
If anyone wants to try their hand at implementing this integration library I can setup a separate repo.

**Reply by thomhurst** [reply]
I've started taking a look myself. Do we need a new repo? I was going to submit a pr here

**Reply by aivascu** [reply]
@thomhurst the repo got a little too big and it's becoming increasingly hard to juggle versions of the .NET framework and upstream dependencies. Currently in v5 everything is waiting for everything else to be done so it can be released. 😅 
I'm currently working on extracting the commit history for a couple libraries from the main repo.

**Reply by thomhurst** [reply]
I've raised a PR into `release/5.0.0`: https://github.com/AutoFixture/AutoFixture/pull/1497

If you want me to move it into its own repo, let me know. Though I'm not sure how it'd reference the core project. Would it be a package reference instead of a project reference?

**Comment by Stepami** [comment]
is there a preview on NuGet?
